pkg/csi/cinder/controllerserver.go

```diff
@@ -345,6 +415,15 @@ func (cs *controllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 		return nil, status.Error(codes.Internal, "Failed to get snapshots")
 	}
 
+	// Verify a backup with the provided name doesn't already exist for this tenant
```

Comment:
the below code is extremely complex and hard to read. I'd suggest to split it into two parts: snapshots only, backups only. The mixing of number if/then/else-if/else-if/else is not an option.